lib/components/event-queue.js

```diff
@@ -0,0 +1,160 @@
+     */
+    push(event, dataReady) {
+        const queue = this._getQueue(event);
+        queue.events.push({
+            dataReady: dataReady
+        });
+    }
+
+    _getQueue(event) {
+        const identifier = this.type === "per_room" ? event.room_id : "none";
+        if (!this._queues[identifier]) {
+            this._queues[identifier] = {
+                events: [],
+                consuming: false
+            };
+        }
+        return this._queues[identifier];
+    }
+
+    /**
+     * Starts consuming the queue.
+     *
+     * As long as events are enqueued they will continue to be consumed.
+     */
+    consume() {
+        Object.keys(this._queues).forEach((identifier) => {
+            if (!this._queues[identifier].consuming) {
+                this._queues[identifier].consuming = true;
+                this._takeNext(identifier);
+            }
+        });
+    }
+
+    _takeNext(identifier) {
+        const events = this._queues[identifier].events;
+        if (events.length === 0) {
+            this._queues[identifier].consuming = false;
+            return;
+        }
+        const entry = events.shift();
+
+        Bluebird.resolve(entry.dataReady).asCallback(this.consumeFn);
+        entry.dataReady.finally(() => this._takeNext(identifier));
+    }
+
+    /**
+     * Factory for EventQueues.
+     *
+     * @param {"none"|"single"|"per_room"} opts.type Type of the queue to create.
+     * @param {consumeCallback} consumeFn Function which is called when an event
+     *     is consumed.
+     * @return {EventQueue} The newly created EventQueue.
+     */
+    static create(opts, consumeFn) {
+        const type = opts.type;
+        /* eslint-disable no-use-before-define */
+        if (type == "single") {
+            return new EventQueueSingle(consumeFn);
+        }
+        if (type == "per_room") {
+            return new EventQueuePerRoom(consumeFn);
+        }
+        if (type == "none") {
+            return new EventQueueNone(consumeFn);
+        }
+        /* eslint-enable no-use-before-define */
+        throw Error(`Invalid EventQueue type '${type}'.`);
+    }
+}
+
+/**
+ * EventQueue for which all events are enqueued in their order of arrival.
+ *
+ * The foremost event is processed as soon as its data is available.
+ */
+class EventQueueSingle extends EventQueue {
+    constructor(consumeFn) {
+        super("single", consumeFn);
+    }
+}
+
+/**
+ * EventQueue for which one queue per room is utilized.
+ *
+ * Events at the head of line are processed as soon as their data is available.
+ */
+class EventQueuePerRoom extends EventQueue {
+    constructor(consumeFn) {
+        super("per_room", consumeFn);
+    }
+}
+
+/**
+ * Dummy EventQueue for which no queue is utilized.
+ *
+ * Every event is handled as soon as its data is available.
+ */
+class EventQueueNone extends EventQueue {
+    constructor(consumeFn) {
+        super("none", consumeFn);
+    }
+
+    push(event, data, promise) {
```

Comment:
as far as I can tell, you never pass a third value into this function anywhere else in the PR which makes it go bang.